external/java/src/main/java/ch/epfl/dedis/skipchain/SkipchainRPC.java

```diff
@@ -74,39 +84,114 @@ public SkipBlock getSkipblock(SkipblockId id) throws CothorityCommunicationExcep
 
         try {
             SkipchainProto.SkipBlock sb = SkipchainProto.SkipBlock.parseFrom(msg);
-            //TODO: add verification that the skipblock is valid by hashing and comparing to the id
+            SkipBlock ret = new SkipBlock(sb);
+
+            // simple verification (we do not check the links, just the signature)
+            if (!ret.verifyForwardSignatures()) {
+                throw new CothorityCommunicationException("invalid forward signatures");
+            }
 
             logger.debug("Got the following skipblock: {}", sb);
             logger.info("Successfully read skipblock");
 
-            return new SkipBlock(sb);
+            return ret;
         } catch (InvalidProtocolBufferException e) {
             throw new CothorityCommunicationException(e);
         }
     }
 
-    public SkipBlock getLatestSkipblock() throws CothorityCommunicationException {
-        SkipchainProto.GetUpdateChain request =
-                SkipchainProto.GetUpdateChain.newBuilder().setLatestID(ByteString.copyFrom(scID.getId())).build();
+    /**
+     * Returns the latest block and verify that the links are correct. The verification is performed from the genesis block.
+     *
+     * @return the latest skipblock
+     * @throws CothorityCommunicationException if something goes wrong with communication
+     * @throws CothorityCryptoException if the verification goes wrong
+     */
+    public SkipBlock getLatestSkipblock() throws CothorityCommunicationException, CothorityCryptoException {
+        return this.getLatestSkipblock(this.scID);
+    }
 
-        ByteString msg = roster.sendMessage("Skipchain/GetUpdateChain",
-                request);
+    /**
+     * Returns the latest block and verify that the links are correct. The verification is performed from trustedLatest.
+     *
+     * @param trustedLatest is the latest block ID that the caller trusts, which serves as the source for verification.
+     * @return the latest skipblock
+     * @throws CothorityCommunicationException if something goes wrong with communication
+     * @throws CothorityCryptoException if the verification goes wrong
+     */
+    public SkipBlock getLatestSkipblock(SkipblockId trustedLatest) throws CothorityCommunicationException, CothorityCryptoException {
+        List<SkipBlock> update = new ArrayList<>();
+        for (;;) {
+            // make the request
+            SkipchainProto.GetUpdateChainReply r2;
+            try {
+                SkipchainProto.GetUpdateChain request =
+                        SkipchainProto.GetUpdateChain.newBuilder()
+                                .setLatestID(ByteString.copyFrom(trustedLatest.getId()))
```

Comment:
A get update chain request with no MaxHeight will traverse the highest links possible (see comments on skipchain/msgs.go, line 97). But the code below seems to assume it is working on level 0 links only?

Reply:
See comment here https://github.com/dedis/cothority/pull/1649#discussion_r247058325